backfill: introduce BackfillExpr

otan opened the pull request:
This is an alternative method to #60310. I kind of prefer #60310, but this
does give us the nice property that rows written during backfill will insert
correctly. The reason I prefer the other is it is less "pollutey".

----


We previously relied on the gateway_region of the schema changer to
dictate what the rows would be populated as. However, in actuality, we
want the region to be the PRIMARY REGION when changing the table to
REGIONAL BY ROW for the first time.

This pattern is unfortunately unsupported by the schema changer right now:

```
BEGIN;
-- Run backfill on all columns to 5.
ALTER TABLE t ADD COLUMN col INT DEFAULT 5.
-- We want any new rows afterwards will default to 10.
-- Unfortunately with the current schema changer world, the data will
all backfill to 10.
ALTER TABLE t ALTER COLUMN col SET DEFAULT 10;
COMMIT
```

To rectify this, we introduce the concept of a `BackfillExpr` on a
ColumnDef, which backfills default expressions separately to default
expressions.

Alternative to #60310.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60476)
<!-- Reviewable:end -->

otan committed:
backfill: introduce BackfillExpr

We previously relied on the gateway_region of the schema changer to
dictate what the rows would be populated as. However, in actuality, we
want the region to be the PRIMARY REGION when changing the table to
REGIONAL BY ROW for the first time.

This pattern is unfortunately unsupported by the schema changer right now:

```
BEGIN;
-- Run backfill on all columns to 5.
ALTER TABLE t ADD COLUMN col INT DEFAULT 5.
-- We want any new rows afterwards will default to 10.
-- Unfortunately with the current schema changer world, the data will
all backfill to 10.
ALTER TABLE t ALTER COLUMN col SET DEFAULT 10;
COMMIT
```

To rectify this, we introduce the concept of a `BackfillExpr` on a
ColumnDef, which backfills default expressions separately to default
expressions.

Release note: None

ajwerner reviewed (commented):
I think I prefer the other one; I'm not eager to further complicate the table descriptor. You?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60476#-:-MTH_8SD-PtULE37f7Up:b-6ejflj)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
agreed, was worth a shot to see how ugly it got though :')
